src/view/position.js

```diff
@@ -243,61 +243,39 @@ export default class Position {
 	 * @returns {module:engine/view/position~PositionRelation}
 	 */
 	compareWith( otherPosition ) {
+		if ( this.root !== otherPosition.root ) {
+			return 'different';
+		}
+
 		if ( this.isEqual( otherPosition ) ) {
 			return 'same';
 		}
 
-		// If positions have same parent.
-		if ( this.parent === otherPosition.parent ) {
-			return this.offset - otherPosition.offset < 0 ? 'before' : 'after';
-		}
+		// Get path from root to position's parent element. "Indexify" the paths (change elements to indices).
+		const thisPath = this.getAncestors().map( element => element.index );
+		const otherPath = otherPosition.getAncestors().map( element => element.index );
```

Comment:
WDYT about using just introduced `Node#getPath()`?

```javascript
const thisPath = this.parent.is( 'node' ) ? this.parent.getPath() : [];
const otherPath = otherPosition.parent.is( 'node' ) ? otherPosition.parent.getPath() : []
```
and shifting `null` from array won't be necessary.